src/cdk-experimental/listbox/listbox.ts

```diff
@@ -6,26 +6,128 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
-import {Directive} from '@angular/core';
+import {
+  AfterContentInit,
+  ContentChildren,
+  Directive,
+  ElementRef,
+  Input,
+  QueryList
+} from '@angular/core';
 
+/**
+ * Directive that applies interaction patterns to an element following the aria role of option.
+ * Typically meant to be placed inside a listbox. Logic handling selection, disabled state, and
+ * value is built in.
+ */
 @Directive({
   selector: '[cdkOption]',
   exportAs: 'cdkOption',
   host: {
     role: 'option',
+    '[attr.aria-selected]': '_selected',
+    '[attr.data-optionid]': '_optionId',
   }
 })
 export class CdkOption {
+  private _selected: boolean | null = null;
+  private _optionId: string;
+
+  /** Whether the option is selected or not */
+  @Input()
+  get selected(): boolean | null {
+    return this._selected;
+  }
+  set selected(value: boolean | null) {
+    this._selected = value;
+  }
+
+  constructor(private _el: ElementRef) {
+  }
 
+  /** Sets the optionId to the given id */
+  setOptionId(id: string): void {
+    this._optionId = id;
+  }
+
+  /** Returns the optionId of this option */
+  getOptionId(): string {
+    return this._optionId;
+  }
+
+  /** Returns an ElementRef of this option */
+  getElementRef(): ElementRef {
+    return this._el;
+  }
 }
 
+let _uniqueIdCounter = 0;
+
+/**
+ * Directive that applies interaction patterns to an element following the aria role of listbox.
+ * Typically CdkOption elements are placed inside the listbox. Logic to handle keyboard navigation,
+ * selection of options, active options, and disabled states is built in.
+ */
 @Directive({
   selector: '[cdkListbox]',
   exportAs: 'cdkListbox',
   host: {
     role: 'listbox',
+    '(click)': 'onClickUpdateSelectedOption($event)'
   }
 })
-export class CdkListbox {
+export class CdkListbox implements AfterContentInit {
 
+  constructor() {
+  }
+
+  /** A query list containing all CdkOption elements within this listbox */
+  @ContentChildren(CdkOption, {descendants: true}) _options: QueryList<CdkOption>;
+
+  /** On click handler of this listbox, updates selected value of clicked option */
+  onClickUpdateSelectedOption($event: MouseEvent) {
+    this._options.toArray().forEach(option => {
+      const optionId = option.getOptionId();
+      if ($event.target instanceof Element &&
+          optionId === $event.target?.getAttribute('data-optionid')) {
+        this._updateSelectedOption(option);
+      }
```

Comment:
Interesting idea. I was planning on doing some performance comparisons and testing this weekend and refactor it afterwards.

Reply:
One problem here: the option element may not actually be the click target; the option may (likely will) contain other elements. This would mean you'd need to walk up the DOM to get the option. You could do something here like:

```ts
updateSelection(event: MouseEvent) {
  const optionElement = getClosestOption(event.target);
  const selectedOption = this._options.toArray().find(o => o.id === optionElement.id);
  ...
}
```

As for attaching the component instance to the DOM- it's not particularly aligned with Angular's development model. The main thing I'd worry about is leaking memory since the DOM element and the component would reference each other.

Reply:
Agreed on the solution being weird in Angular (although I'd also argue Angular is a little deficient for handling this particular use case), but I don't think memory leaking is an issue as long as the reference is unset in ngOnDestroy.

You might be remembering the circular reference issue from IE6/IE7 (https://stackoverflow.com/questions/10092619/precise-explanation-of-javascript-dom-circular-reference-issue), but this particular one hasn't been an issue for a while :)

Reply:
I guess I'm showing my age

Reply:
Another method is to have the option listen for the click and toggle its selected state on click. Then it uses an injected reference to its parent listbox to trigger the listbox's selectionChange event emitter. However I think this may go against the "read not write" relationship the option should have with its parent listbox.